mechanisms/BuildModules.cmake

```diff
@@ -56,3 +56,77 @@ function(build_modules)
+
+  cmake_parse_arguments(MK_CAT "" "NAME;SOURCES;OUTPUT" "MECHS" ${ARGN})
+  set(MK_CAT_OUT_DIR "${CMAKE_CURRENT_BINARY_DIR}/generated/${MK_CAT_NAME}")
+
+  # Need to set ARB_WITH_EXTERNAL_MODCC *and* modcc
+  set(external_modcc)
+  if(ARB_WITH_EXTERNAL_MODCC)
+    set(external_modcc MODCC ${modcc})
+  endif()
+
+  message(NOTICE "Catalogue name:       ${MK_CAT_NAME}")
+  message(NOTICE "Catalogue mechanisms: ${MK_CAT_MECHS}")
+  message(NOTICE "Catalogue sources:    ${MK_CAT_SOURCES}")
+  message(NOTICE "Catalogue output:     ${MK_CAT_OUT_DIR}")
+
+  file(MAKE_DIRECTORY "${MK_CAT_OUT_DIR}")
+
+  if (NOT TARGET build_all_mods)
+    add_custom_target(build_all_mods)
+  endif()
+
+  build_modules(
+    ${MK_CAT_MECHS}
+    SOURCE_DIR "${MK_CAT_SOURCES}"
+    DEST_DIR "${MK_CAT_OUT_DIR}"
+    ${external_modcc} # NB: expands to 'MODCC <binary>' to add an optional argument
+    MODCC_FLAGS -t cpu -t gpu ${ARB_MODCC_FLAGS} -N arb::${MK_CAT_NAME}_catalogue
+    GENERATES .hpp _cpu.cpp _gpu.cpp _gpu.cu
+    TARGET build_catalogue_${MK_CAT_NAME}_mods)
+
+  set(catalogue_${MK_CAT_NAME}_source ${CMAKE_CURRENT_BINARY_DIR}/${MK_CAT_NAME}_catalogue.cpp)
+  set(catalogue_${MK_CAT_NAME}_options -A arbor -I ${MK_CAT_OUT_DIR} -o ${catalogue_${MK_CAT_NAME}_source} -B multicore -C ${MK_CAT_NAME} -N arb::${MK_CAT_NAME}_catalogue)
+  if(ARB_WITH_GPU)
+    list(APPEND catalogue_${MK_CAT_NAME}_options -B gpu)
+  endif()
+
+  add_custom_command(
+    OUTPUT ${catalogue_${MK_CAT_NAME}_source}
+    COMMAND ${CMAKE_CURRENT_SOURCE_DIR}/generate_catalogue ${catalogue_${MK_CAT_NAME}_options} ${MK_CAT_MECHS}
+    COMMENT "Building catalogue ${MK_CAT_NAME}"
+    DEPENDS generate_catalogue)
+
+  add_custom_target(${MK_CAT_NAME}_catalogue_cpp_target DEPENDS ${catalogue_${MK_CAT_NAME}_source})
+  add_dependencies(build_catalogue_${MK_CAT_NAME}_mods ${MK_CAT_NAME}_catalogue_cpp_target)
+  add_dependencies(build_all_mods build_catalogue_${MK_CAT_NAME}_mods)
+
+  foreach(mech ${MK_CAT_MECHS})
+    list(APPEND catalogue_${MK_CAT_NAME}_source ${MK_CAT_OUT_DIR}/${mech}_cpu.cpp)
+    if(ARB_WITH_GPU)
+      list(APPEND catalogue_${MK_CAT_NAME}_source ${MK_CAT_OUT_DIR}/${mech}_gpu.cpp ${MK_CAT_OUT_DIR}/${mech}_gpu.cu)
+    endif()
+  endforeach()
+  set(${MK_CAT_OUTPUT} ${catalogue_${MK_CAT_NAME}_source} PARENT_SCOPE)
+
+  set_source_files_properties(${catalogue_${MK_CAT_NAME}_source} COMPILE_FLAGS ${ARB_CXXOPT_ARCH})
+  add_library(${MK_CAT_NAME}-catalogue SHARED ${catalogue_${MK_CAT_NAME}_source})
+  target_compile_definitions(${MK_CAT_NAME}-catalogue PUBLIC STANDALONE=1)
+  set_target_properties(${MK_CAT_NAME}-catalogue
+    PROPERTIES
+    SUFFIX ".so"
+    PREFIX "lib"
+    CXX_STANDARD 17)
+  target_include_directories(${MK_CAT_NAME}-catalogue PUBLIC "${ARB_SOURCE_DIR}/arbor/")
+
+  if(${CMAKE_PROJECT_NAME} STREQUAL arbor)
+    target_link_libraries(${MK_CAT_NAME}-catalogue PRIVATE arbor)
+  else()
+    target_link_libraries(${MK_CAT_NAME}-catalogue PRIVATE arbor::arbor)
```

Comment:
This (and above) will attempt to link in the non PIC libarbor.a (which is what we get when we're not asking for python libs), and this will cause a link failure. What it does need though is `arbor::arbor-public-headers`.

Reply:
Ok, that means, I will add an entry to the package file to record whether PIC was enabled and error out if not.

Reply:
There's no need ... the shared object doesn't need to be linked against arbor. It should be fine having a non-PIC `libarbor.a` dynamically load a PIC `foo-catalogue.so`.

Reply:
But it does need the con/destructors of `mechanism` which are in `libarbor.a`, linking the catalogue fails like this
```
  ....
  "typeinfo for arb::multicore::mechanism", referenced from:
      typeinfo for arb::cat_catalogue::mechanism_cpu_dummy in dummy_cpu.cpp.o
  "vtable for arb::multicore::mechanism", referenced from:
      arb::multicore::mechanism::mechanism() in dummy_cpu.cpp.o
      arb::multicore::mechanism::~mechanism() in dummy_cpu.cpp.o
  NOTE: a missing vtable usually means the first non-inline virtual member function has no definition.
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```
This is also the result of the missing ABI, as we are deriving from a base class defined in the arbor library.

Reply:
Gah. It's never clear to me what does and does not work in this direction. An unresolved function call would be fine, for example (presuming no visibility shenanigans).

Reply:
A bit of googling indicates this might be resolvable, tbh, if every virtual method in the parent classes has a definition within `libarbor.a`.

Reply:
We determined offline that we need to link against `libarbor.a`.